File: src/main/java/net/lavaclient/client/module/ModuleManager.java
package net.lavaclient.client.module;

import net.lavaclient.client.event.EventTarget;
import net.lavaclient.client.event.events.UpdateEvent;
import net.lavaclient.client.module.modules.combat.Criticals;
import net.lavaclient.client.module.modules.combat.KillAura;
import net.lavaclient.client.module.modules.combat.Velocity;
import net.lavaclient.client.module.modules.misc.AntiBot;
import net.lavaclient.client.module.modules.movement.Flight;
import net.lavaclient.client.module.modules.movement.Speed;
import net.lavaclient.client.module.modules.player.NoFall;
import net.lavaclient.client.module.modules.render.HUD;
import org.lwjgl.input.Keyboard;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Manages all client modules
 */
public class ModuleManager {
    private final List<Module> modules = new ArrayList<>();
    
    /**
     * Constructor initializes all modules
     */
    public ModuleManager() {
        // Register all modules
        
        // Combat
        registerModule(new KillAura());
        registerModule(new Velocity());
        registerModule(new Criticals());
        registerModule(new AntiBot());
        
        // Movement
        registerModule(new Speed());
        registerModule(new Flight());
        
        // Player
        registerModule(new NoFall());
        
        // Render
        registerModule(new HUD());
        
        // Set default keybinds
        getModule(KillAura.class).setKeyBind(Keyboard.KEY_R);
        getModule(Velocity.class).setKeyBind(Keyboard.KEY_V);
        getModule(Speed.class).setKeyBind(Keyboard.KEY_X);
        getModule(Flight.class).setKeyBind(Keyboard.KEY_F);
        getModule(NoFall.class).setKeyBind(Keyboard.KEY_N);
        
        // HUD is always enabled
        getModule(HUD.class).setState(true);
        
        System.out.println("Registered " + modules.size() + " modules");
    }
    
    /**
     * Registers a module
     * @param module The module to register
     */
    public void registerModule(Module module) {
        modules.add(module);
    }
    
    /**
     * Gets a module by class
     * @param clazz The module class
     * @param <T> The module type
     * @return The module instance
     */
    @SuppressWarnings("unchecked")
    public <T extends Module> T getModule(Class<T> clazz) {
        for (Module module : modules) {
            if (module.getClass() == clazz) {
                return (T) module;
            }
        }
        
        return null;
    }
    
    /**
     * Gets a module by name
     * @param name The module name
     * @return The module instance
     */
    public Module getModule(String name) {
        for (Module module : modules) {
            if (module.getName().equalsIgnoreCase(name)) {
                return module;
            }
        }
        
        return null;
    }
    
    /**
     * Gets all modules
     * @return List of all modules
     */
    public List<Module> getModules() {
        return modules;
    }
    
    /**
     * Gets modules by category
     * @param category The category
     * @return List of modules in the category
     */
    public List<Module> getModulesByCategory(Module.Category category) {
        return modules.stream()
                .filter(module -> module.getCategory() == category)
                .collect(Collectors.toList());
    }
    
    /**
     * Gets enabled modules
     * @return List of enabled modules
     */
    public List<Module> getEnabledModules() {
        return modules.stream()
                .filter(Module::getState)
                .collect(Collectors.toList());
    }
    
    /**
     * Update event target for key binds
     * @param event Update event
     */
    @EventTarget
    public void onUpdate(UpdateEvent event) {
        // Check key binds
        for (Module module : modules) {
            if (module.getKeyBind() != 0 && Keyboard.isKeyDown(module.getKeyBind())) {
                module.toggle();
            }
        }
    }
}

File: src/main/kotlin/net/lavaclient/client/module/ModuleManager.kt
/*
 * This file is part of LavaClient.
 * https://github.com/lavaclient/client
 */
package net.lavaclient.client.module

import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.lavaclient.client.LavaClient
import net.lavaclient.client.module.impl.client.ClickGuiModule

/**
 * Manages all modules
 */
class ModuleManager {
    // List of all modules
    private val modules = mutableListOf<Module>()
    
    /**
     * Initialize the module manager
     */
    fun init() {
        // Register modules
        registerModule(ClickGuiModule())
        
        // Log the number of registered modules
        LavaClient.INSTANCE.info("Registered ${modules.size} modules")
        
        // Register tick event for module updates
        ClientTickEvents.END_CLIENT_TICK.register {
            // Call onTick for enabled modules
            modules.filter { it.enabled }.forEach { it.onTick() }
        }
    }
    
    /**
     * Register a module
     *
     * @param module The module to register
     */
    private fun registerModule(module: Module) {
        modules.add(module)
    }
    
    /**
     * Get a module by name
     *
     * @param name The name of the module
     * @return The module or null if not found
     */
    fun getModule(name: String): Module? {
        return modules.find { it.name.equals(name, ignoreCase = true) }
    }
    
    /**
     * Get all modules
     *
     * @return All modules
     */
    fun getModules(): List<Module> {
        return modules
    }
    
    /**
     * Get modules by category
     *
     * @param category The category
     * @return Modules in the category
     */
    fun getModulesByCategory(category: Category): List<Module> {
        return modules.filter { it.category == category }
    }
}